templated-this.md

```diff
@@ -1,151 +1,315 @@
+towards making methods as powerful as inline friend functions, with the only
+difference being the call syntax.
+
+One really cool implication of this is that the `this` parameter would be
+move-constructed in the case where the object is an _rvalue_, allowing you to
+treat chained builder methods that return a new object uniformly.
+
+*Example:*
+
+```cpp
+class string_builder {
+  std::string s;
+
+  operator std::string (T this) {
+    return std::move(s);
+  }
+  string_builder operator*(T this, int n) {
+    assert(n > 0);
+    
+    s.reserve(s.size() * n);
+    auto const size = s.size();
+    for (auto i = 0; i < n; ++i) {
+      s.append(s, 0, size);
+    }
+    return this;
+  }
+  string_builder bop(T this) {
+    s.append("bop");
+    return this;
+  }
+};
+
+// this is optimally efficient as far as allocations go
+std::string const x = (string_builder{{"asdf"}} * 5).bop();
+```
+
+Of course, implementing this example with templated `this` methods would have
+been more efficient due to just having fewer objects, but we got rid of all
+references in the program!
+
+
+### `this` as a reference
+
+This paper turns `this` into a reference on an opt-in basis, which is in line
+with the existing guidance that never-null pointers should be references if at
+all possible, and in this case, it is possible.
+
+We believe there would be no confusion, as in all cases, the value category of
+`this` is stated plain as day in the parameter list, which is on the very same
+screen.
+
+Teaching also becomes easier, as the meaning of what a 'const method' is becomes
+more obvious to students.
+
+One can always obtain the address of the object by taking the address of the
+`this`.
+
+
+### unification with `inline friend` functions
+
+This proposal also makes `this` far less special. In fact, it completely unifies
+`inline friend` functions and class methods, with the differences being:
+
+- the calling convention
+- methods can be virtual
+
+Basically, if the first parameter is called `this`, one can parse and
+instantiate the declaration with exactly the same rules as an `inline friend`,
+except with a calling convention for methods.
+
+
+### `virtual`
+
+We are not entirely sure how `virtual` and value calling conventions would work,
+so perhaps we can disallow that use-case, if the pass-by-value signatures are
+kept in.
+
+Implications for lambdas
+------------------------
+
+Generic lambdas, should they take an `auto&& this` parameter, work according to
+existing rewriting rules: the `auto&& this` is turned into a "universal
+reference" and deduced as if it were inside a
+`template <typename T> auto operator()(T&& this) -> <return type> {...}`.)
+
+### Interplays with capturing `[this]` and `[*this]`
+
+As an explicit parameter, it shadows members of the closure. That said, using
+`this` inside the lambda is useless for anything other than `decltype(this)`,
+since the closure members have no defined names.
```

Comment:
Not sure exactly what you mean here, maybe we need a code example.

Reply:
This is straight from Marshall: the members of the closure object have no defined names, and the closure object has no defined layout or even a requirement to store all the members that were captured (it can totally decide to store some in registers) if the compiler can prove things.
If you capture `[this]`, then it's visible in your function, because it's treated basically like a member would be - except here, the parameter `this` would shadow it.

Reply:
I see what you mean now. It sounded odd to say "closure members have no defined names" when clearly I can give them names in the capture list. But those names are not connected to the closure object is what we're saying. IOW, given:
```cpp
int x;
auto l = [x] (auto&& this) { ... };
```
Inside `l` I can obviously refer to `x` but there's no requirement that the hypothetical `this.x` exists in memory, and it is not reachable by that name.

Reply:
Reprhased. We need to find a solution for that for sure though, even if it's a new std::forward_like<decltype(this)>(m). Honestly that would be useful regardless, perhaps another paper?